### DISCUSSION TITLE
Get data passed from _document to _app nextjs

### DISCUSSION THREAD

**Discussion by nhattran998** [vercel/next.js#35271] [Help]
Hi everyone, I currently use Material UI + NextJs and I have a problem with the media query which rehydrate not correct on the server-side(pretty obvious).

So I read the docs - https://mui.com/components/use-media-query/#server-side-rendering, then I need to figure out the way to pass the device type( `const deviceType = parser(req.headers['user-agent']).device.type || 'desktop'` ) which I get from `_document.tsx` then pass to `_app.tsx`.

```typescript
static async getInitialProps(ctx: DocumentContext) {
    const userAgent = ctx.req
      ? ctx.req.headers['user-agent']
      : navigator.userAgent;

    const deviceType = UaParser(userAgent).device.type || 'desktop';

    return {
     .....,
      deviceType
     };
  }
```
    
   So I try to pass the `deviceType` to the `Main`(I think I pass this and the _app` can get that but NOT)
    
```typescript
 render() {
    const { deviceType } = this.props;
    return (
      <Html>
        <body>
          <Main deviceType={deviceType} /> // Not worked
          <NextScript />
        </body>
      </Html>
    );
  }
```
  
  In `_app.tsx`, I need to receive the `deviceType` I pass from server-side, but I cannot know how to do this. So anyone can help me 🥲.

```typescript
  function MyApp({
  Component,
  pageProps
}: MyAppProps) {
 // I need deviceType variable here 😆 to control correct the media query of Material UI
}
```

**Comment by FINDarkside** [comment]
I don't think passing state from _document to _app is possible, because `Document.getInitialProps` is called after `App.getInitialProps`. However, what you could do is to compute `deviceType` inside middleware, set it in header like `x-device-type`. Then you can access the header inside `_app`:
```
export default function App({ Component, pageProps, deviceType }) {
  // deviceType prop will only be available during SSR
  const [cachedDeviceType] = useState(deviceType);
  return <Component {...pageProps} />;
}

App.getInitialProps = async ({ ctx }) => {
  if (ctx.res) {
    const deviceType = ctx.res.getHeader("x-device-type");
    return {
      deviceType,
    };
  }
  return {};
};
```
